test-loop-tests/src/tests/catching_up.rs

```diff
@@ -0,0 +1,515 @@
+}
+
+// Use another height to send txs.
+#[test]
+fn ultra_slow_test_catchup_random_single_part_sync_height_9() {
+    test_catchup_random_single_part_sync_common(false, false, 9)
+}
+
+fn test_catchup_random_single_part_sync_common(skip_24: bool, non_zero: bool, send_tx_height: u64) {
+    init_test_logger();
+
+    let validators: Vec<Vec<AccountId>> = [
+        vec!["test1.1", "test1.2", "test1.3", "test1.4"],
+        vec!["test2.1", "test2.2", "test2.3", "test2.4"],
+        vec![
+            "test3.1", "test3.2", "test3.3", "test3.4", "test3.5", "test3.6", "test3.7", "test3.8",
+        ],
+    ]
+    .iter()
+    .map(|vec| vec.iter().map(|account| account.parse().unwrap()).collect())
+    .collect();
+    let seats: NumSeats = 8;
+
+    let stake = ONE_NEAR;
+    let mut runner = RotatingValidatorsRunner::new(stake, validators);
+
+    let test_accounts: Vec<AccountId> =
+        (0..16).map(|i| format!("test_account{i}").parse().unwrap()).collect_vec();
+
+    let all_accounts =
+        runner.all_validators_accounts().iter().chain(test_accounts.iter()).cloned().collect_vec();
+
+    let num_shards = 4;
+    let epoch_length = 8;
+    let genesis = TestLoopBuilder::new_genesis_builder()
+        .epoch_length(epoch_length)
+        .validators_spec(runner.genesis_validators_spec(seats, seats, seats))
+        .add_user_accounts_simple(&all_accounts, stake)
+        .shard_layout(ShardLayout::multi_shard(num_shards, 3))
+        .build();
+
+    let epoch_config_store = TestEpochConfigBuilder::from_genesis(&genesis)
+        .minimum_validators_per_shard(2)
+        .build_store_for_genesis_protocol_version();
+
+    let block_prod_time = Duration::milliseconds(6000);
+    let mut env = TestLoopBuilder::new()
+        .genesis(genesis)
+        .clients(runner.all_validators_accounts())
+        .epoch_config_store(epoch_config_store)
+        .config_modifier(move |config, _| {
+            config.min_block_production_delay = block_prod_time;
+            config.max_block_production_delay = 3 * block_prod_time;
+            config.max_block_wait_delay = 3 * block_prod_time;
+        })
+        .build()
+        .warmup();
+
+    for node_datas in &env.node_datas {
+        let peer_actor_handle = node_datas.peer_manager_sender.actor_handle();
+        let peer_actor = env.test_loop.data.get_mut(&peer_actor_handle);
+        peer_actor.register_override_handler(Box::new(move |request| -> Option<NetworkRequests> {
+            if let NetworkRequests::PartialEncodedChunkMessage { partial_encoded_chunk, .. } =
+                &request
+            {
+                if skip_24 {
+                    if partial_encoded_chunk.header.height_created() == 23
+                        || partial_encoded_chunk.header.height_created() == 24
+                    {
+                        return None;
+                    }
+                }
+            }
+            Some(request)
```

Comment:
nit: If trivial can you move it to helper function?

Reply:
It's not complicated, but I don't see why would it improve things. I think it make it a bit harder to read the test.